src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs

```diff
@@ -356,16 +356,43 @@ protected void CheckDocumentIsOpen(DocumentId documentId)
         }
 
         protected internal void OnDocumentOpened(DocumentId documentId, SourceTextContainer textContainer, bool isCurrentContext = true)
+            => OnDocumentOpened(documentId, textContainer, isCurrentContext, requireDocumentPresentAndClosed: true);
+
+        internal void TryOnDocumentOpened(DocumentId documentId, SourceTextContainer textContainer, bool isCurrentContext)
+            => OnDocumentOpened(documentId, textContainer, isCurrentContext, requireDocumentPresentAndClosed: false);
+
+        internal void OnDocumentOpened(DocumentId documentId, SourceTextContainer textContainer, bool isCurrentContext, bool requireDocumentPresentAndClosed)
         {
             SetCurrentSolution(
                 static (oldSolution, data) =>
                 {
-                    var (@this, documentId, _, _) = data;
-
-                    CheckDocumentIsInSolution(oldSolution, documentId);
-                    @this.CheckDocumentIsClosed(documentId);
+                    var (@this, documentId, _, _, requireDocumentPresentAndClosed) = data;
 
                     var oldDocument = oldSolution.GetRequiredDocument(documentId);
+                    if (oldDocument is null)
+                    {
+                        // Didn't have a document.  Throw if required.  Bail out gracefully if not.
+                        if (requireDocumentPresentAndClosed)
+                        {
+                            throw new ArgumentException(string.Format(
+                                WorkspacesResources._0_is_not_part_of_the_workspace,
+                                oldSolution.Workspace.GetDocumentName(documentId)));
+                        }
+                        else
+                        {
+                            return oldSolution;
+                        }
+                    }
+
+                    if (@this.IsDocumentOpen(documentId))
+                    {
+                        // Document was already open.  Throw if required.  Bail out gracefully if not.
+                        if (requireDocumentPresentAndClosed)
+                            @this.CheckDocumentIsClosed(documentId);
+                        else
+                            return oldSolution;
+                    }
```

Comment:
similar logic to what was already done in OnAnyDocumentTextChanged.  Slightly augmented in that it is doing two checks (doc is present, and it's in teh expected open/close state).